### PR TITLE
Verify datum type matches column type

### PR DESCRIPTION
**Pull request by thschroeter**
Giving #2033 a try. Seems to be easier than #629.

### CODE REVIEW

File: sql/table.go [hunk 1]
```diff
@@ -300,29 +300,45 @@ func encodeSecondaryIndexes(tableID structured.ID, indexes []structured.IndexDes
...
+		if v, ok := val.(parser.DString); ok {
+			return string(v), nil
+		}
+		valTypeOk = false
+	}
+	if !valTypeOk {
+		return nil, fmt.Errorf("column type %s doesn't match value type %T", col.Type.Kind, val)
 	}
-	return nil, fmt.Errorf("Unsupported type: %T", val)
+	return nil, fmt.Errorf("unsupported type: %T", val)
```
Comment (tamird) [inline sql/table.go]: you can do without `valTypeOk` by putting this in a `default` arm of the match above and then returning the "column type doesn't match" error here.

Comment (petermattis) [inline sql/table.go]: You can get rid of `valTypeOk` if you add a `default` case statement to the switch.

**Comment by tamird** [comment]
needs a test

File: sql/table.go [hunk 2]
```diff
@@ -300,29 +300,45 @@ func encodeSecondaryIndexes(tableID structured.ID, indexes []structured.IndexDes
...
 	// case structured.ColumnType_TIMESTAMP:
 	case structured.ColumnType_CHAR, structured.ColumnType_TEXT, structured.ColumnType_BLOB:
-		return string(val.(parser.DString)), nil
+		if v, ok := val.(parser.DString); ok {
+			return string(v), nil
+		}
+		valTypeOk = false
+	}
+	if !valTypeOk {
+		return nil, fmt.Errorf("column type %s doesn't match value type %T", col.Type.Kind, val)
```
Comment (petermattis) [inline sql/table.go]: Change the `%T` to a `%s` and use `val.Type()` for the argument. You'll also need to change the `val` parameter type to `parser.Datum` which it should have been all along.

**Comment by petermattis** [comment]
Agreed with Tamir that this needs a few tests. Should be easy to add something to `testdata/insert`. 

Thanks for fixing this.

**Comment by thschroeter** [comment]
Addressed comments. Since we need to return at the end of the function, I left one error message at the end, without the default branch.

File: sql/table.go [hunk 3]
```diff
@@ -300,29 +300,36 @@ func encodeSecondaryIndexes(tableID structured.ID, indexes []structured.IndexDes
...
-func prepareVal(col structured.ColumnDescriptor, val parser.Expr) (interface{}, error) {
+func prepareVal(col structured.ColumnDescriptor, val parser.Datum) (interface{}, error) {
 	if val == parser.DNull {
 		return nil, nil
 	}
 
 	switch col.Type.Kind {
 	case structured.ColumnType_BIT:
-		return bool(val.(parser.DBool)), nil
+		if v, ok := val.(parser.DBool); ok {
```
Comment (petermattis) [inline sql/table.go]: Just noticed that this was incorrect before your change. `BIT` types should be stored as ints. Mind fleshing out the test cases.

**Comment by tamird** [comment]
@thschroeter why not have the default arm? It should return an error (or panic) indicating that the switch is an incomplete match.

File: sql/testdata/insert [hunk 4]
```diff
@@ -92,6 +92,23 @@ INSERT INTO kv3 VALUES ('a', NULL)
 statement error null value in column "v" violates not-null constraint
 INSERT INTO kv3 (k) VALUES ('a')
 
+
```
Comment (tamird) [inline sql/testdata/insert]: remove this space?

File: sql/table.go (hunk 3 above)
Comment (petermattis) [inline sql/table.go]: Oh, and you can combine this case statement with `ColumnType_INT`:

```
  case structured.ColumnType_BIT, structured.ColumnType_INT:
```

File: sql/testdata/insert (hunk 4 above)
Comment (thschroeter) [inline sql/testdata/insert]: yes

File: sql/table.go [hunk 5]
```diff
@@ -300,29 +300,36 @@ func encodeSecondaryIndexes(tableID structured.ID, indexes []structured.IndexDes
...
 	// case structured.ColumnType_TIME:
 	// case structured.ColumnType_TIMESTAMP:
 	case structured.ColumnType_CHAR, structured.ColumnType_TEXT, structured.ColumnType_BLOB:
-		return string(val.(parser.DString)), nil
+		if v, ok := val.(parser.DString); ok {
+			return string(v), nil
+		}
 	}
-	return nil, fmt.Errorf("Unsupported type: %T", val)
+	return nil, fmt.Errorf("column type %s doesn't match value type %s", col.Type.Kind, val.Type())
```
Comment (tamird) [inline sql/table.go]: we should log the column name as well

**Comment by thschroeter** [comment]
Will try to add different error messages for unsupported column type and column type not equal value type.

File: sql/table.go (hunk 3 above)
Comment (thschroeter) [inline sql/table.go]: ok, will return `int64` for `BIT` and `INT`.

**Comment by thschroeter** [comment]
Addressed comments.

**Comment by tamird** [comment]
LGTM can you squash?

**Comment by thschroeter** [comment]
done

**Comment by thschroeter** [comment]
fixing the test, sorry

**Comment by thschroeter** [comment]
The test fails after rebase, I think. `create table t (k int primary key);
insert into t values('a')` seems to incorrectly return ok now, running `select * from t` panics.

**Comment by tamird** [comment]
@thschroeter looking

**Comment by tamird** [comment]
@thschroeter found the regression. PR in a moment.

**Comment by tamird** [comment]
@thschroeter you should be good to go after a rebase. Sorry about that!

**Comment by thschroeter** [comment]
Rebased. Thank you.

**Comment by tamird** [comment]
Thanks!

**Comment by thschroeter** [comment]
Closes #2033. Forgot to mention this in the commit message, so the issue is still open.

**Pull request merged**